File: agent/tools/email_tool.py
"""
Email sending tool wrapper
"""
from .base_tool import BaseTool
from utils.api_client import BackendAPIClient
from typing import Dict, Any


class EmailTool(BaseTool):
    """Tool for sending booking confirmation emails"""
    
    def __init__(self):
        super().__init__(
            name="send-email",
            description="Send booking confirmation email"
        )
        self.api_client = BackendAPIClient()
    
    async def execute(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """
        Send email confirmation
        Args:
            params: {'bookingId': 'booking_id'}
        Returns:
            Email sending status
        """
        print(f"🔧 [EMAIL_TOOL] Executing with params: {params}")
        error = self.validate_params(params, ['bookingId'])
        if error:
            print(f"❌ [EMAIL_TOOL] Validation failed: {error}")
            return {'success': False, 'error': error}
        
        try:
            result = await self.api_client.call_tool('send-email', params)
            print(f"✅ [EMAIL_TOOL] Email sent successfully")
            return result
        except Exception as e:
            print(f"❌ [EMAIL_TOOL] Error: {str(e)}")
            import traceback
            traceback.print_exc()
            return {
                'success': False,
                'error': str(e)
            }

File: backend/tools/emailTool.js
/**
 * Email sending tool (standalone, separate from booking creation)
 */
const { sendBookingConfirmation } = require("../services/emailService");
const Booking = require("../models/Booking");

/**
 * Send email confirmation
 * @param {Object} params - Tool parameters
 * @param {string} params.bookingId - Booking ID to send confirmation for
 * @returns {Promise<Object>} Email sending status
 */
const execute = async (params) => {
  console.log(`🔧 [EMAIL_TOOL] Starting email send`);
  console.log(`📦 [EMAIL_TOOL] Params:`, JSON.stringify(params, null, 2));

  try {
    const { bookingId } = params;

    if (!bookingId) {
      const error = "Booking ID parameter is required";
      console.error(`❌ [EMAIL_TOOL] ${error}`);
      throw new Error(error);
    }

    // Find the booking
    console.log(`🔍 [EMAIL_TOOL] Finding booking with ID: ${bookingId}`);
    const booking = await Booking.findById(bookingId);
    if (!booking) {
      const error = "Booking not found";
      console.error(`❌ [EMAIL_TOOL] ${error}`);
      throw new Error(error);
    }

    // Send email confirmation
    console.log(
      `📧 [EMAIL_TOOL] Sending confirmation email to ${
        booking.customerEmail || "restaurant"
      }`
    );
    await sendBookingConfirmation(booking);

    console.log(`✅ [EMAIL_TOOL] Email sent successfully`);

    return {
      success: true,
      data: {
        bookingId,
        emailSent: true,
        customerEmail: booking.customerEmail || null,
        restaurantEmail: process.env.RESTAURANT_EMAIL || null,
      },
    };
  } catch (error) {
    // Log error but return structured response
    console.error(`❌ [EMAIL_TOOL] Error:`, error.message);
    console.error("Stack trace:", error.stack);
    return {
      success: false,
      error: error.message || "Failed to send email",
    };
  }
};

module.exports = {
  name: "send-email",
  description: "Send booking confirmation email",
  execute,
};
